overlord/fdestate/fdemgr.go

```diff
@@ -412,6 +414,55 @@ func GetRecoveryKey(st *state.State, keyID string) (rkey keys.RecoveryKey, err e
 	return mgr.getRecoveryKey(keyID)
 }
 
+// devPathByContainerRole returned a map from container role to corresponding
+// device path. If containerRoles is nil, all encrypted containers are returned.
+// If any encrypted container role from containerRoles does not exist, an error
+// is returned.
+func (m *FDEManager) devPathByContainerRole(containerRoles []string) (map[string]string, error) {
+	containers, err := m.GetEncryptedContainers()
+	if err != nil {
+		return nil, err
+	}
+
+	allContainers := len(containerRoles) == 0
+	devPathByContainerRole := make(map[string]string, len(containers))
+	for _, container := range containers {
+		if allContainers || strutil.ListContains(containerRoles, container.ContainerRole()) {
+			devPathByContainerRole[container.ContainerRole()] = container.DevPath()
+		}
+	}
+
+	if !allContainers {
+		for _, containerRole := range containerRoles {
+			if _, ok := devPathByContainerRole[containerRole]; !ok {
+				return nil, fmt.Errorf("encrypted container role %q does not exist", containerRole)
+			}
+		}
+	}
+
+	return devPathByContainerRole, nil
+}
+
+// CheckRecoveryKey tests that the specified recovery key unlocks the
+// specified container roles. If containerRoles is nil, the recovery
+// key will be tested against all container roles.
+//
+// The state must be locked by the caller.
+func (m *FDEManager) CheckRecoveryKey(rkey keys.RecoveryKey, containerRoles []string) error {
+	devPathByContainerRole, err := m.devPathByContainerRole(containerRoles)
```

Comment:
I am not sure why you had to return a map here. Unless you plan to use `devPathByContainerRole` somewhere else?

Reply:
I am actually using it here https://github.com/canonical/snapd/pull/15557/files#diff-5527da57557ae929e605cc7c905e7edee3b61711671c3c9ef40dce5a1e21ff2bR406, but starting to feel it doesn't add value, I will try removing it and see if it is cleaner this way.

Reply:
I refactored the code to move away from it, thanks!